dep0/lib/03_typecheck/src/delta_unfold.cpp

```diff
@@ -1,16 +1,330 @@
+    // eg `f(x)` for some global function `f`, but not everywhere, eg `return f`.
+    // The reason for this is that we are interested in delta-unfolding only as
+    // a way to extend beta normalization as far as possible.
+    return false;
+}
+
+bool delta_unfold(environment_t const& env, context_t const& ctx, expr_t::app_t& app)
+{
+    if (auto const global = std::get_if<expr_t::global_t>(&app.func.get().value))
         if (auto const func_def = std::get_if<func_def_t>(env[*global]))
         {
-            x.value = func_def->value;
+            app.func.get().value = func_def->value;
             return true;
         }
+    if (delta_unfold(env, ctx, app.func.get()))
+        return true;
+    for (auto& arg: app.args)
+        if (delta_unfold(env, ctx, arg))
+            return true;
     return false;
 }
 
+bool delta_unfold(environment_t const& env, context_t const& ctx, expr_t::abs_t& abs)
+{
+    auto ctx2 = ctx.extend();
+    for (auto& arg: abs.args)
+    {
+        if (delta_unfold(env, ctx2, arg.type))
+            return true;
+        if (arg.var)
+        {
+            auto const inserted = ctx2.try_emplace(*arg.var, arg.properties.origin, make_legal_expr(arg.type, *arg.var));
+            assert(inserted.has_value());
+        }
+    }
+    if (delta_unfold(env, ctx2, abs.ret_type.get()))
+        return true;
+    return delta_unfold(env, ctx2, abs.body);
+}
+
+bool delta_unfold(environment_t const& env, context_t const& ctx, expr_t::pi_t& pi)
+{
+    auto ctx2 = ctx.extend();
+    for (auto& arg: pi.args)
+    {
+        if (delta_unfold(env, ctx2, arg.type))
+            return true;
+        if (arg.var)
+        {
+            auto const inserted = ctx2.try_emplace(*arg.var, arg.properties.origin, make_legal_expr(arg.type, *arg.var));
+            assert(inserted.has_value());
+        }
+    }
+    return delta_unfold(env, ctx2, pi.ret_type.get());
+}
+
+
+bool delta_unfold(environment_t const& env, context_t const& ctx, expr_t::init_list_t& init_list)
+{
+    for (auto& v: init_list.values)
+        if (delta_unfold(env, ctx, v))
+            return true;
+    return false;
+}
+
+bool delta_unfold(environment_t const& env, context_t const& ctx, expr_t::subscript_t& subscript)
+{
+    return delta_unfold(env, ctx, subscript.array.get()) or delta_unfold(env, ctx, subscript.index.get());
+}
+
+} // namespace impl
+
+bool delta_unfold(environment_t const& env, context_t const& ctx, body_t& body)
+{
+    for (auto& s: body.stmts)
+        if (impl::delta_unfold(env, ctx, s))
+            return true;
+    return false;
+}
+
+bool delta_unfold(environment_t const& env, context_t const& ctx, expr_t& expr)
+{
+    if (auto const type = std::get_if<expr_t>(&expr.properties.sort.get()))
+        if (delta_unfold(env, ctx, *type))
+            return true;
+    // for boolean_expr_t, relation_expr_t and arith_expr_t, prefer primitive reduction over further unfolding
```

Comment:
```suggestion
    // for boolean_expr_t, relation_expr_t,  etc, prefer primitive reduction over further unfolding
```

Reply:
But am still undecided on whether this should be in beta reduction instead

Reply:
For arithmetic expressions we cannot use beta reduction because it might depend on the environment